### PR TITLE
Mailbox 296 : EXPUNGE should not generate Cassandra timeouts for 5.000 messages

### PR DESCRIPTION
**Pull request by chibenwa**

### CODE REVIEW

File: server/container/util-java8/src/main/java/org/apache/james/util/streams/JamesCollectors.java [hunk 1]
```diff
@@ -19,26 +19,63 @@
...
+
+        private ChunkCollector(int chunkSize) {
+            Preconditions.checkArgument(chunkSize > 0, "ChunkSize should be strictly positive");
+            this.chunkSize = chunkSize;
+            this.counter = new AtomicInteger(-1);
+        }
+
+        @Override
+        public Supplier<Multimap<Integer, D>> supplier() {
+            return () -> Multimaps.synchronizedListMultimap(ArrayListMultimap.<Integer, D>create());
```
Comment (mbaechler) [inline server/container/util-java8/src/main/java/org/apache/james/util/streams/JamesCollectors.java]: synchronized ? I don't think a collector needs to be thread-safe, did you read that somewhere ?

Comment (mbaechler) [inline server/container/util-java8/src/main/java/org/apache/james/util/streams/JamesCollectors.java]: a list of list should really be enough : you create a list and put elements until size == chunksize and then create a new one, etc. Append all list to a main list and you're done, right ?

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java [hunk 2]
```diff
@@ -220,14 +220,11 @@ public MessageUid findFirstUnseenMessageUid(Mailbox mailbox) throws MailboxExcep
...
         CassandraId mailboxId = (CassandraId) mailbox.getMailboxId();
 
-        return FluentFutureStream.of(deletedMessageDAO.retrieveDeletedMessage(mailboxId, messageRange))
-            .completableFuture()
-            .thenApply(stream -> stream.collect(JamesCollectors.chunker(EXPUNGE_BATCH_SIZE)))
-            .thenCompose(chunkedExpungedUids ->
-                CompletableFutureUtil.chainAll(chunkedExpungedUids,
-                    uidChunk ->  expungeUidChunk(mailboxId, uidChunk)))
-            .thenApply(s -> s.flatMap(i -> i))
+        return deletedMessageDAO.retrieveDeletedMessage(mailboxId, messageRange)
```
Comment (mbaechler) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java]: I don't understand the change : what I see is a join right after the first read, is that what changed ? What was the problem then ?

Comment (chibenwa) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java]: If I use chainAll, I timeout on the original read.

What is a weird issue.

Not using it solved it. I observed similar behaviour on SetFlags.

File: server/container/util-java8/src/main/java/org/apache/james/util/streams/JamesCollectors.java (hunk 1 above)
Comment (chibenwa) [inline server/container/util-java8/src/main/java/org/apache/james/util/streams/JamesCollectors.java]: Ok, will remove concurrency  support on https://github.com/linagora/james-project/pull/806

Comment (chibenwa) [inline server/container/util-java8/src/main/java/org/apache/james/util/streams/JamesCollectors.java]: We have an object that supports list creation for us. Why avoiding it? 

Why bother with last list selection? Last list switch? Multimap avoids these troubles.

Comment (chibenwa) [inline server/container/util-java8/src/main/java/org/apache/james/util/streams/JamesCollectors.java]: And merging 2 multimaps is way easier than merging 2 List of List by the way.

Thus I don't consider this library uses "overkill"

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java (hunk 2 above)
Comment (mbaechler) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java]: too much parallelism ?

Comment (chibenwa) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java]: The way I understand it is *that we may not read Cassandra result fast enough, thus generating  a timeout* even if that explanation seems strange to me...

Comment (chibenwa) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java]: Maybe it's due to interference between paging (of 5000) and slow processing....

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**